DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/main.cpp

```diff
@@ -0,0 +1,500 @@
+  unsigned int crc_h, count_h;
+
+  // track timing information in ms
+  std::vector<double> time_ms(runs);
+
+  // input and output data pointers on the device using USM device allocations
+  unsigned char *in, *out;
+
+  // the number of decompressed bytes (the number of bytes in 'out')
+  // returned by Consumer kernel
+  int* decompressed_count;
+
+  // the GZIP header data (see gzip_header_data.hpp)
+  GzipHeaderData* hdr_data;
+
+  // the GZIP footer data, where 'count' is the expected number of bytes
+  // in the uncompressed file
+  int *crc, *count;
+
+  bool passed = true;
+
+  try {
+    // allocate memory on the device
+    if ((in = malloc_device<unsigned char>(in_count, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'in'\n";
+      std::terminate();
+    }
+    if ((out = malloc_device<unsigned char>(out_count_padded, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'out'\n";
+      std::terminate();
+    }
+    if ((decompressed_count = malloc_device<int>(1, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'decompressed_count'\n";
+      std::terminate();
+    }
+    if ((hdr_data = malloc_device<GzipHeaderData>(1, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'hdr_data'\n";
+      std::terminate();
+    }
+    if ((crc = malloc_device<int>(1, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'crc'\n";
+      std::terminate();
+    }
+    if ((count = malloc_device<int>(1, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'count'\n";
+      std::terminate();
+    }
+
+    // copy the input data to the device memory and wait for the copy to finish
+    q.memcpy(in, in_bytes.data(), in_count * sizeof(unsigned char)).wait();
+
+    std::cout << "Decompressing '" << f_in << "' " << runs
+              << ((runs == 1) ? " time" : " times") << std::endl;
+
+    // run the design multiple times to increase the accuracy of the timing
+    for (int i = 0; i < runs; i++) {
+      std::cout << "Launching kernels for run " << i << std::endl;
+
+      auto producer_event = SubmitProducer(q, in_count, in);
+      auto consumer_event =
+          SubmitConsumer(q, out_count_padded, out, decompressed_count);
+
+      auto gzip_decompress_events =
+          SubmitGzipDecompressKernels<InPipe, OutPipe, kLiteralsPerCycle>(
+              q, in_count, hdr_data, crc, count);
+
+      auto start = high_resolution_clock::now();
+      producer_event.wait();
+      consumer_event.wait();
+      auto end = high_resolution_clock::now();
+
+      // wait for the decompression kernels to finish
+      for (auto& e : gzip_decompress_events) {
+        e.wait();
+      }
+
+      std::cout << "All kernels have finished for run " << i << std::endl;
+
+      // duration in milliseconds
+      time_ms[i] = duration<double, std::milli>(end - start).count();
```

Comment:
Aren't there SYCL runtime functions that we're supposed to use to 'properly' measure kernel execution times?

Reply:
Yes. Don't like them, and std::chrono is just as accurate ...  our SYCL timing is ... not great.